File: java/CutRibbon189A.java
package com.nicholasmoreles.cpsolutions;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

/**
 * 
 * @author Nicholas Moreles
 */
public class CutRibbon189A {
	private static final int NUM_COINS = 3;
	
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		int number = Integer.parseInt(st.nextToken());
		int[] coins = new int[NUM_COINS];
		for (int i = 0; i < NUM_COINS; i++) {
			coins[i] = Integer.parseInt(st.nextToken());
		}
		int[] maxCoins = new int[number + 1];
		Arrays.fill(maxCoins, -1);
		maxCoins[0] = 0;
		for (int i = 1; i <= number; i++) {
			for (int coin : coins) {
				final int prev = i - coin;
				if (prev >= 0 && maxCoins[prev] >= maxCoins[i] && maxCoins[prev] != -1) {
					maxCoins[i] = maxCoins[prev] + 1;
				}
			}
		}
		System.out.println(maxCoins[number]);
	}
}

File: c++/CutRibbon189A.cc
#include <iostream>
#include <vector>

using std::cin;
using std::cout;
using std::vector;

/**
 * Author: Nicholas Moreles
 */
int main() {
  const int NUM_COINS = 3;
  int number;
  int coins[NUM_COINS] = {0};

  cin >> number;
  for (int i = 0; i < NUM_COINS; ++i) {
    cin >> coins[i];
  }

  vector<int> max_coins;
  max_coins.push_back(0);
  for (int i = 1; i <= number; ++i) {
    max_coins.push_back(-1);
    for (int j = 0; j < NUM_COINS; ++j) {
      const int coin = coins[j];
      const int prev = i - coin;
      if (prev >= 0 && max_coins[prev] >= max_coins[i]
            && max_coins[prev] != -1) {
        max_coins[i] = max_coins[prev] + 1;
      }  
    }
  }

  cout << max_coins[number];

  return 0;
}
